src/Autofac/Core/Activators/Reflection/MostParametersConstructorSelector.cs

```diff
@@ -42,13 +42,15 @@ public class MostParametersConstructorSelector : IConstructorSelector
         /// <param name="parameters">Parameters to the instance being resolved.</param>
         /// <returns>The best constructor.</returns>
         /// <exception cref='DependencyResolutionException'>A single unambiguous match could not be chosen.</exception>
-        public ConstructorParameterBinding SelectConstructorBinding(ConstructorParameterBinding[] constructorBindings, IEnumerable<Parameter> parameters)
+        public BoundConstructor SelectConstructorBinding(BoundConstructor[] constructorBindings, IEnumerable<Parameter> parameters)
```

Comment:
I'm somewhat wrapped around the axle a bit on the combination of:

- `ReflectionActivator.ActivateInstance` -> `ReflectionActivator.GetValidConstructorBindings`
- `ReflectionActivator.ActivateInstance` -> `ReflectionActivator.ConstructorSelector.SelectConstructorBinding` -> `MostParametersConstructorSelector.SelectConstructorBinding`

My assumption is that the majority case is that folks are using our default constructor selector, which is `MostParametersConstructorSelector`. Given that and the amount of LINQ there is to calculate each `BoundConstructor`, its number of arguments, I'm curious if there might be some pre-calculation that could be done. For example, if `BoundConstructor` had a `ConstructorParameterLength` field with that already calculated, that'd save on the LINQ and allocations on every instantiation... and wouldn't _hurt_ to have if you didn't need it in your custom selector.

That is, I wonder if we can somewhat optimize for the common use case a little.

I also see a sort of path where there's LINQ/Enumerable and ToArray at various points in the path. Like, `GetValidConstructorBindings` creates an array of `BoundConstructor`, then if there's any that can't be instantiated, it creates _another_ array, this time filtered; which then gets passed to `SelectConstructorBinding` which will create yet a _third_ array (ostensibly of one element)... I'm curious what your thoughts are on that. I know there was a point a while back where we switched several things from `IEnumerable<T>` to arrays to avoid some allocations of iterators and things (#919). Is there anything else we should be doing to help with that?

Reply:
Yeah, the `MostParametersConstructorSelector` does need a big overhaul, it does a **lot** of allocating. The reason I've left it out of the initial changes was just because I consider the overwhelmingly most likely use-case is having a single constructor, in which case the method returns immediately.

I'm wondering if this comes back to my early thoughts on static analysis, where rather than iterate over each constructor and try to resolve each parameter multiple times, we just determine the set of all dependencies on the object, and then try to resolve those, then yielding the set of constructors based on what we can provide.

That would require more work at container build time, but it may have legs; I'll have a think.

Reply:
As for memory allocation, I have considered if we could just pass 'all' the constructors into the selector, including invalid ones, and let the selector ignore the bad ones. If we improve the performance of `MostParametersConstructorSelector`, it may be worth considering.

It would be a breaking change for anyone with a custom selector, but I can't imagine that is a large number of people.

Reply:
> That would require more work at container build time, but it may have legs; I'll have a think.

Haha, that didn't take long; `Parameter` instances receive the `ParameterInfo` instance, so could make decisions beyond the type (like attributes on the argument), so we can't just build a graph of unique dependencies, because we don't know what logic a custom `Parameter` might apply.

I'll just fix the `MostParametersConstructorSelector`.

Reply:
Also, we need a benchmark for components with multiple valid constructors. I'll raise a PR to put that into develop.